### ISSUE TITLE
Requests are executed only synchronously

### ISSUE THREAD

**Issue by segabor** [swift-server/async-http-client#129]
I'm writing a simple HTTP test client that communicates with Apache Tomcat server on localhost. It operates with simple HEAD requests.
If I execute request the suggested way it does effectively nothing
```swift
let httpClient = HTTPClient(eventLoopGroupProvider: .createNew)

...

var request = try HTTPClient.Request(url: "http://localhost:8080/...", method: .HEAD)
request.headers.add(name: "User-Agent", value: "Swift HTTPClient")
request.body = .none
httpClient.execute(request: request).whenComplete() { result in
    switch result {
    case .failure(let error):
        // process error
        print("ERR \(error)")
    case .success(let response):
        if response.status == .ok {
            // handle response
            print("RESP: \(response)")
        } else {
            // handle remote error
            print(response.status)
        }
    }
}
```
Callback never called.

But if I wait for the result it will complete.
```swift
if let result = try? httpClient.execute(request: request).wait() {
    print(result.status)
}
// prints ok
```
What am I doing wrong? Is it possible that event loop is not running?

I'm using Swift 5.1.2 in Arch Linux.

**Comment by segabor** [comment]
Ok, so if I add the deferred shutdown right after `HTTPClient` construction, `whenComplete()` will be called and it reports "ERR shutdown" back.

**Comment by segabor** [comment]
Well, it's getting weird. I tried the same sample code on my MacBook Pro (Catalina + Xcode latest) connecting to Tomcat remotely. Now I get the following error:

```
NIOConnectionError(host: "csihuhu", port: 8080, dnsAError: nil, dnsAAAAError: nil, connectionErrors: [NIO.SingleConnectionFailure(target: [IPv4]csihuhu/192.168.221.43:8080, error: NIO.ChannelError.ioOnClosedChannel)])
ERR NIOConnectionError(host: "csihuhu", port: 8080, dnsAError: nil, dnsAAAAError: nil, connectionErrors: [NIO.SingleConnectionFailure(target: [IPv4]csihuhu/192.168.221.43:8080, error: NIO.ChannelError.ioOnClosedChannel)])
```
I definitely doing it wrong, no clue what.

**Comment by weissi** [comment]
CC @artemredkin

**Comment by weissi** [comment]
@segabor are you maybe shutting down  the `httpClient` before the requests arrive?

**Comment by segabor** [comment]
No, not at all. I was just following the code template described in README.md to write my simple http client code. It does nothing special but sending some HEAD request to my Tomcat server.

**Comment by weissi** [comment]
@segabor when are you calling `syncShutdown`?

**Comment by segabor** [comment]
It's being embedded in the `defer` statement at the beginning of the main source.

```swift
let httpClient = HTTPClient(eventLoopGroupProvider: .createNew)
defer {
    try? httpClient.syncShutdown()
}
```

**Comment by weissi** [comment]
@segabor and you're waiting for the asynchronous responses? `defer` will execute at the end of the current scope

**Comment by segabor** [comment]
Got it. I should wait until all requests are processed at the end of main() code. That makes sense.
How can I do that?

**Comment by t089** [comment]
Shouldn't shutdown already gracefully wait for inflight events to finish before it stops the event loop?

**Comment by weissi** [comment]
@segabor ideally you’d only start one HTTPClient at the beginning in your app and shut it down when your app exits

**Comment by weissi** [comment]
@t089 it should wait or cancel them, yes. Mind filing a bug?

**Comment by segabor** [comment]
> @segabor ideally you’d only start one HTTPClient at the beginning in your app and shut it down when your app exits

This is just a command-line app. After having a good sleep on this I realized what was going on. My tool finished way before handlers get called. What I need in my case is good old linear, blocking web calls.
So no issue here, closing this ticket.

**Comment by weissi** [comment]
@segabor awesome! Yes, just as you say, in a cmdline utility that doesn't need to do multiple calls in parallel, just making them blocking by `.wait()`ing on them and executing them one after the other sounds like the winning plan.

**Comment by weissi** [comment]
@t089 filed #131 to clarify (and improve) the semantics of shutdown.
